src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -6876,50 +6897,51 @@ public override BoundNode VisitLambda(BoundLambda node)
             // conversion case, need to investigate others
             if (!_disableNullabilityAnalysis)
             {
-                var bag = new DiagnosticBag();
-                VisitLambda(node, delegateTypeOpt: null, bag);
-                bag.Free();
+                VisitLambda(node, delegateTypeOpt: null, disableDiagnostics: true);
             }
             SetNotNullResult(node);
             return null;
         }
 
-        private void VisitLambda(BoundLambda node, NamedTypeSymbol delegateTypeOpt, DiagnosticBag diagnostics, VariableState initialState = null)
+        private void VisitLambda(BoundLambda node, NamedTypeSymbol? delegateTypeOpt, Optional<LocalState> initialState = default, bool disableDiagnostics = false)
         {
             Debug.Assert(delegateTypeOpt?.IsDelegateType() != false);
-            var analyzedNullabilityMap = _analyzedNullabilityMapOpt;
-            var snapshotBuilder = _snapshotBuilderOpt;
-            if (_disableNullabilityAnalysis)
-            {
-                analyzedNullabilityMap = null;
-                snapshotBuilder = null;
-            }
-            else if (_snapshotBuilderOpt is object && delegateTypeOpt is object && !node.UnboundLambda.HasExplicitlyTypedParameterList)
+
+            var delegateInvokeMethod = delegateTypeOpt?.DelegateInvokeMethod;
+            bool useDelegateInvokeParameterTypes = UseDelegateInvokeParameterTypes(node, delegateInvokeMethod);
+            if (useDelegateInvokeParameterTypes && _snapshotBuilderOpt is object)
             {
                 SetUpdatedSymbol(node, node.Symbol, delegateTypeOpt);
             }
 
-            Analyze(
```

Comment:
By no longer calling this analyze overload, we're not entering a new scope for the lambda, so speculation will be broken. We need to restore the calls to `snapshotManager?.EnterNewWalker` with the new symbol.